File: boj/1012.cpp
#include<bits/stdc++.h>
using namespace std;

int adj[51][51];
bool visited[51][51];
int N,M,K;
int dx[4] = {-1,0,1,0};
int dy[4] = {0,-1,0,1};
int cnt;

void dfs(int y, int x){

	visited[y][x] = true;
	
	for(int i=0; i<4; i++){
		
		int nextDy = y + dy[i];
		int nextDx = x + dx[i];
		
		if(nextDy<0 || nextDx <0  || nextDy >= N || nextDx >= M) continue;
		
		if(!visited[nextDy][nextDx] && adj[nextDy][nextDx] ==1){
			dfs(nextDy,nextDx);
		}
	}
}

int main(void){
	ios::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);
	
	int T;
	cin >> T;
	
	while(T--){
		cin >> M >> N >> K;
		cnt = 0;
		memset(adj,0,sizeof(adj));
		memset(visited,0,sizeof(visited));
		while(K--){
			int X,Y;
			cin >> X >> Y;
			adj[Y][X] = 1;
		}
		
		for(int i=0; i<N; i++){
			for(int j=0; j<M; j++){
				if(adj[i][j] == 1 && !visited[i][j]){
					dfs(i,j);
					cnt++;
				}
			}
		}
		cout << cnt << '\n';
	}
	return 0;
}

File: boj/1012.py
import sys
sys.setrecursionlimit(10**5)

T= int(input())

dy = [-1,0,1,0]
dx = [0,-1,0,1]

def dfs(y,x):
    for i in range(4):
        nextX = x + dx[i]
        nextY = y + dy[i]
        if nextX < 0 or nextY <0 or nextX >= M or nextY >= N or visited[nextY][nextX]==True :
            continue
        if MAP[nextY][nextX]==1:
            visited[nextY][nextX] = True
            dfs(nextY,nextX)

for i in range(T):
    M,N,K = map(int,input().split())
    MAP=[ [0] * M for _ in range(N)]
    visited = [ [False ]* M for _ in range(N)]
    cnt =0
    for _ in range(K):
        X,Y=map(int,input().split())
        MAP[Y][X] = 1

    for y in range(N):
        for x in range(M):
            if MAP[y][x] == 1 and not visited[y][x]:
                visited[y][x] = True
                cnt+=1
                dfs(y,x)
    print(cnt)